src/webcore/executor.rs

```diff
@@ -0,0 +1,194 @@
+            // Clear `is_queued` flag so that it will re-queue if poll calls task.notify()
+            self.is_queued.set( false );
+
+            if spawn_future.poll_future_notify( &&EventLoop, self as *const _ as usize ) == Ok( Async::NotReady ) {
+                // Future was not ready, so put it back
+                *spawn = Some( spawn_future );
+            }
+        }
+    }
+
+    fn notify( task: Rc< SpawnedTask > ) {
+        // If not already queued
+        if !task.is_queued.replace( true ) {
+            EventLoop.push_task(task);
+        }
+    }
+}
+
+// A proxy for the javascript event loop.
+struct EventLoop;
+
+// There's only one thread, but this lets us tell the compiler that we
+// don't need a `Sync` bound, and also gives us lazy initialization.
+thread_local! {
+    static EVENT_LOOP_INNER: EventLoopInner = EventLoopInner::new();
+}
+
+impl EventLoop {
+    fn drain(&self) {
+        EVENT_LOOP_INNER.with(EventLoopInner::drain)
+    }
+    fn push_task(&self, task: Rc< SpawnedTask >) {
+        EVENT_LOOP_INNER.with(|inner| inner.push_task(task))
+    }
+}
+
+// State relating to the javascript event loop. Only one instance ever exists.
+struct EventLoopInner {
+    // Avoid unnecessary allocation and interop by keeping a local
+    // queue of pending tasks.
+    microtask_queue: RefCell< VecDeque< Rc< SpawnedTask > > >,
+    waker: Reference,
+}
+
+// Not strictly necessary, but may become relevant in the future
+impl Drop for EventLoopInner {
+    #[inline]
+    fn drop( &mut self ) {
+        js! { @(no_return)
+            @{&self.waker}.drop();
+        }
+    }
+}
+
+impl EventLoopInner {
+    // Initializes the event loop. Only called once.
+    fn new() -> Self {
+        EventLoopInner {
+            microtask_queue: RefCell::new(VecDeque::with_capacity(INITIAL_QUEUE_CAPACITY)),
+            waker: js!(
+                var callback = @{|| EventLoop.drain()};
+                var wrapper = function() { callback() };
+
+                // Modern browsers can use `MutationObserver` which allows
+                // us to schedule a micro-task without allocating a promise.
+                // https://dom.spec.whatwg.org/#notify-mutation-observers
+                if ( typeof MutationObserver === "function" ) {
+                    var node = document.createTextNode( "0" );
+                    var state = false;
+
+                    new MutationObserver( wrapper ).observe( node, { characterData: true } );
+
+                    function nextTick() {
+                        state = !state;
+                        node.data = ( state ? "1" : "0" );
+                    }
+                    nextTick.drop = callback.drop;
+
+                    return nextTick;
+
+                // Node.js and other environments
+                } else {
+                    var promise = Promise.resolve( null );
+
+                    function nextTick( value ) {
+                        promise.then( wrapper );
```

Comment:
When the `EventLoopInner` is dropped while the Promise is pending, it will cause a runtime panic. So in that situation it shouldn't call the `callback`

Reply:
Right now the `EventLoopInner` is never dropped (since the main thread never exits) but I will adjust this anyway.